### PR TITLE
Major compaction 1st edition

### PR DESCRIPTION
**Pull request by wayblink**

### CODE REVIEW

File: configs/milvus.yaml [hunk 1]
```diff
@@ -435,6 +435,22 @@ dataCoord:
     maxParallelTaskNum: 10 # max parallel compaction task number
     indexBasedCompaction: true
 
+    major:
```
Comment (wayblink) [inline configs/milvus.yaml:438]: About the terminology here：we have major and L2:

I think L2 request user to understand the concept of segment type and relationships between different compaction types and segment types. Here compaction is actually L1 + L2 -> L2,  and we will add L2 single compaction, L2 merge compaction in the future. So I think L2 may cause more confusion. 

For major compaction, we will split one major compaction into multi sub compaction tasks to handle large dataset and handoff. It is a little weird to call these subTask as `major`. We developers should notice that.

I prefer `major` than `L2` for easier user understanding. And another proposal is `clusteringCompaction`.
`clustering` is more accurate and echos the concept of `clusteringKey`

Final compaction type will be like:
  Level0DeleteCompaction
  SingleCompaction:  add support for L2
  MinorCompaction: add support for L2
  ClusteringCompaction(MajorCompaction)： L1+L2 -> L2

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+    // random sampling to get train data
+    std::vector<int32_t> idx(file_paths.size());
+    std::iota(idx.begin(), idx.end(), 0);
+    std::shuffle(idx.begin(), idx.end(), std::mt19937());
+    int selected_size = 0;
+    auto buf = std::unique_ptr<T[]>(new T[train_size / sizeof(T)]);
+    int64_t offset = 0;
+
+    for (int i = 0; i < file_paths.size(); i++) {
+        if (selected_size < train_size &&
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:89]: the logic here is correct, but very confusing

for (int i = 0; i < file_paths.size(); i++) {
    auto cur_size = std::min(train_size - selected_size, file_sizes[idx[i]]);
    local_chunk_manager->Read(
        file_paths[idx[i]], 0, reinterpret_cast<char*>(buf.get()) + offset, cur_size);
    selected_size += cur_size;
    offset += cur_size;

    if (selected_size >= train_size) {
        break;
    }
}

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+                                 const std::vector<uint64_t>& file_sizes,
+                                 uint64_t train_size,
+                                 uint64_t total_size) {
+    auto local_chunk_manager =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    // train data fits in memory, read by sequence and generate centroids and id_mapping in one pass
+    if (train_size >= total_size) {
+        auto buf = std::unique_ptr<T[]>(new T[total_size / sizeof(T)]);
+        int64_t offset = 0;
+        for (int i = 0; i < file_paths.size(); i++) {
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:70]: file read can be paralleled

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    auto data_num = data_size / sizeof(T) / dim;
+
+    auto train_num = train_size.value() / sizeof(T) / dim;
+
+    // make train num equal to data num
+    if (train_num >= data_num) {
+        train_num = data_num;
+    }
+    auto train_size_new = train_num * dim * sizeof(T);
+    auto buf = Sample(data_files, offsets, train_size_new, data_size);
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:185]: need a log for sample time

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+    auto res =
+        knowhere::kmeans::ClusteringMajorCompaction<T>(*dataset, num_clusters);
+    if (!res.has_value()) {
+        PanicInfo(ErrorCode::UnexpectedError,
+                  fmt::format("failed to kmeans train: {}: {}",
+                              KnowhereStatusString(res.error()),
+                              res.what()));
+    }
+    dataset.reset();  // release train data
+    auto centroids = reinterpret_cast<const T*>(res.value()->GetTensor());
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:200]: Move the rest of the function to another function

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+    if (local_chunk_manager->Exist(output_path)) {
+        LOG_INFO("path {} already exists, delete it", output_path);
+        local_chunk_manager->RemoveDir(output_path);
+    }
+    local_chunk_manager->CreateDir(output_path);
+    std::string centroid_stats_path = output_path + "centroids";
+    result_files_.emplace_back(centroid_stats_path);
+    WritePBFile(stats, centroid_stats_path);
+
+    auto compute_num_in_centroid = [&](const uint32_t* centroid_id_mapping,
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:232]: all the below should be another function

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+               "segment size is empty when major compaction");
+    auto train_size =
+        milvus::index::GetValueFromConfig<int64_t>(config_, "train_size");
+    AssertInfo(train_size.has_value(),
+               "train size is empty when major compaction");
+
+    std::vector<std::string> data_files;
+    std::vector<uint64_t> offsets;
+    uint32_t dim = 0;
+    auto data_size = file_manager_->CacheCompactionRawDataToDisk(
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:170]: when did cached file cleaned?

We need to check everywhere about the cache file. It should be cleaned if the analysis cleaned or failed

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+        uint64_t start = 0;
+        uint64_t gather_size = 0;
+        // choose half of train size as a group to compute centroids
+        uint64_t group_size = train_size_new / 2;
+        std::vector<int64_t> gather_segment_id;
+        for (auto it = insert_files.value().begin();
+             it != insert_files.value().end();
+             it++) {
+            gather_segment_id.emplace_back(it->first);
+            gather_size += offsets[i];
```
Comment (xiaofan-luan) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:278]: is there a special reason this has to be grouped?

how large WritePBFile could be? I'm assuming we can write everything into one file.

If stats is less than 10GB, we can parallelly read all the files and concurrently write into a same stats , and finally writePBFile to psersistent

File: internal/distributed/indexnode/service.go [hunk 9]
```diff
@@ -289,6 +289,18 @@ func (s *Server) GetMetrics(ctx context.Context, request *milvuspb.GetMetricsReq
 	return s.indexnode.GetMetrics(ctx, request)
 }
 
+func (s *Server) Analysis(ctx context.Context, request *indexpb.AnalysisRequest) (*commonpb.Status, error) {
```
Comment (xiaofan-luan) [inline internal/distributed/indexnode/service.go:292]: use CreateJob and it should support both AnalysisRequest and IndexRequest, something like
message CreateJobRequestV2 {
   AnalysisRequest
   IndexRequest
}

File: internal/querynodev2/delegator/segment_pruner.go [hunk 10]
```diff
@@ -15,12 +15,12 @@ import (
...
-	"github.com/milvus-io/milvus/internal/util/typeutil"
 	"github.com/milvus-io/milvus/pkg/common"
 	"github.com/milvus-io/milvus/pkg/log"
 	"github.com/milvus-io/milvus/pkg/util/distance"
 	"github.com/milvus-io/milvus/pkg/util/funcutil"
 	"github.com/milvus-io/milvus/pkg/util/merr"
+	"github.com/milvus-io/milvus/pkg/util/typeutil"
 )
 
 const defaultFilterRatio float64 = 0.5
```
Comment (xiaofan-luan) [inline internal/querynodev2/delegator/segment_pruner.go:26]: need monitering metrics about pruned ratio here.

File: internal/querynodev2/delegator/segment_pruner.go [hunk 11]
```diff
@@ -152,6 +164,7 @@ func FilterSegmentsByVector(partitionStats *storage.PartitionStatsSnapshot,
 					}
 					// currently, we only support float vector and only one center one segment
 					if disErr != nil {
+						log.Error("calculate distance error", zap.Error(disErr))
```
Comment (xiaofan-luan) [inline internal/querynodev2/delegator/segment_pruner.go:167]: Below this line 

		switch searchReq.GetMetricType() {
		case distance.L2:
			sort.SliceStable(segmentsToSearch, func(i, j int) bool {
				return segmentsToSearch[i].distance < segmentsToSearch[j].distance
			})
		case distance.IP, distance.COSINE:
			sort.SliceStable(segmentsToSearch, func(i, j int) bool {
				return segmentsToSearch[i].distance > segmentsToSearch[j].distance
			})
		}


what if search request don't have metric type?

File: internal/querynodev2/delegator/segment_pruner.go (hunk 10 above)
Comment (xiaofan-luan) [inline internal/querynodev2/delegator/segment_pruner.go:26]: filter ratio might not be good enough, we need a lookup table here.
If segment <= 16 , do not filtering
[17, 32],  segment * 0.5
[33, 256],  16 + (segment - 32) * 0.2
[256, 1024], 60 +  (segment - 256) * 0.05
[1024, more], 98 + (segment - 1024) * 0.01

Comment (xiaofan-luan) [reply internal/querynodev2/delegator/segment_pruner.go:26]: and we should make sure this is tunable

Comment (xiaofan-luan) [reply internal/querynodev2/delegator/segment_pruner.go:26]: need to test recall on multiple dataset as well

Comment (xiaofan-luan) [reply internal/querynodev2/delegator/segment_pruner.go:26]: ideally we expect segment size to be 100+

File: internal/metastore/kv/datacoord/kv_catalog.go [hunk 12]
```diff
@@ -815,3 +815,78 @@ func (kc *Catalog) GcConfirm(ctx context.Context, collectionID, partitionID type
...
+		if err != nil {
+			return nil, err
+		}
+		tasks = append(tasks, model.UnmarshalAnalysisTask(task))
+	}
+	return tasks, nil
+}
+
+func (kc *Catalog) SaveAnalysisTask(ctx context.Context, task *model.AnalysisTask) error {
+	key := buildAnalysisTaskKey(task.TaskID)
```
Comment (jaime0815) [inline internal/metastore/kv/datacoord/kv_catalog.go:875]: suggest adding d bid and collection id into the key which scan keys is easier in multiple db and collections scene

**Comment by xiaofan-luan** [comment]
we can not assume we can use local disk for caching(Currently I thought It should be more like pure memory + remote storage mode).

All of the IOs need to be done through chunkManager. and We should carefully design the cleanup path

File: configs/milvus.yaml [hunk 13]
```diff
@@ -435,6 +435,23 @@ dataCoord:
     maxParallelTaskNum: 10 # max parallel compaction task number
     indexBasedCompaction: true
 
+    clustering:
```
Comment (czs007) [inline configs/milvus.yaml:438]: need to talk about the configuratin

File: internal/core/src/indexbuilder/KmeansMajorCompaction.h [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        Config& config,
+        const storage::FileManagerContext& file_manager_context =
+            storage::FileManagerContext());
+
+    void
+    Train() override;
+
+    BinarySet
+    Upload() override;
+
```
Comment (liliu-z) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.h:41]: Why we need to separate these two parts?
Also `Train` & `Upload` are not suitable APIs for a `Job`, a `Job` can only do run.

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+        train_num = data_num;
+    }
+    auto train_size_new = train_num * dim * sizeof(T);
+    auto buf = Sample(data_files, offsets, train_size_new, data_size);
+    auto dataset = GenDataset(train_num, dim, buf.release());
+
+    // get num of clusters by whole data size / segment size
+    int num_clusters =
+        DIV_ROUND_UP(data_size, segment_size.value());
+    auto res =
```
[the remark this answers is not in the document]
Comment (liliu-z) [reply internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:191]: Talked offline. The definition of this class is a clustering job handling s3 pulling, disk manipulation, knowhere calling, etc.
It is Ok for this class to have only one simple API, (But of course not a huge one, some smaller private functions are needed). And Knowhere's API need to be more functional

File: internal/core/src/indexbuilder/KmeansMajorCompaction.cpp [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+        throw SegcoreError(FileOpenFailed, err_msg.str());
+    }
+
+    outfile.seekp(0, std::ios::beg);
+    if (!message.SerializeToOstream(&outfile)) {
+        std::stringstream err_msg;
+        err_msg << "Error: write local file '" << file_path << " failed, "
+                << strerror(errno);
+        throw SegcoreError(FileWriteFailed, err_msg.str());
+    }
```
Comment (liliu-z) [inline internal/core/src/indexbuilder/KmeansMajorCompaction.cpp:143]: Need to close the stream

**Review by liliu-z — COMMENTED** [review]
Two many disk operations without error handling, destruction mixed with clustering job logic. Let's create a new disk util file to handle all these

**Pull request merged**